ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/handlers/batch/JsonRpcBatchExecutor.java

```diff
@@ -0,0 +1,181 @@
+   * @param tracer The tracer used for monitoring and debugging purposes.
+   * @param ctx The context of the routing, containing information about the HTTP request and
+   *     response.
+   * @param jsonRpcConfiguration The configuration for JSON RPC operations, including the maximum
+   *     batch size for resource-intensive requests.
+   */
+  public JsonRpcBatchExecutor(
+      final JsonRpcExecutor jsonRpcExecutor,
+      final Tracer tracer,
+      final RoutingContext ctx,
+      final JsonRpcConfiguration jsonRpcConfiguration) {
+    this.jsonRpcExecutor = jsonRpcExecutor;
+    this.tracer = tracer;
+    this.ctx = ctx;
+    this.jsonRpcConfiguration = jsonRpcConfiguration;
+  }
+
+  /**
+   * Executes a batch of JSON RPC requests.
+   *
+   * @param rpcRequestBatch A JsonArray containing individual JSON RPC requests.
+   * @return A List of JsonRpcResponse objects, each corresponding to the result of an individual
+   *     request.
+   */
+  public List<JsonRpcResponse> executeRpcRequestBatch(final JsonArray rpcRequestBatch) {
+    final List<JsonRpcResponse> rpcResponseList = new ArrayList<>();
+
+    for (int i = 0; i < rpcRequestBatch.size(); i++) {
+      try {
+        final JsonObject individualRpcRequest = rpcRequestBatch.getJsonObject(i);
+        rpcResponseList.add(processRpcRequest(individualRpcRequest));
+      } catch (final ClassCastException exception) {
+        // In case of invalid request format, add an error response
+        rpcResponseList.add(new JsonRpcErrorResponse(null, INVALID_REQUEST));
+      }
+    }
+
+    return rpcResponseList;
+  }
+
+  /**
+   * Processes an individual RPC request. Determines if the request is resource-intensive and
+   * handles it accordingly.
+   *
+   * @param rpcRequest JsonObject representing the RPC request.
+   * @return A JsonRpcResponse corresponding to the result of the request.
+   */
+  private JsonRpcResponse processRpcRequest(final JsonObject rpcRequest) {
+    if (isResourceIntensiveRequest(rpcRequest)) {
+      return handleResourceIntensiveRequest(rpcRequest);
+    } else {
+      return executeRpcRequest(rpcRequest);
+    }
+  }
+
+  /**
+   * Processes a resource-intensive RPC request. Checks if the request exceeds the limit per batch
+   * and handles it accordingly.
+   *
+   * @param rpcRequest JsonObject representing the RPC request.
+   * @return A JsonRpcResponse corresponding to the result of the request.
+   */
+  private JsonRpcResponse handleResourceIntensiveRequest(final JsonObject rpcRequest) {
+    if (canProcessResourceIntensiveRequest(rpcRequest)) {
+      resourceIntensiveRequestsCounter++;
+      return executeRpcRequest(rpcRequest);
+    } else {
+      // If the request is resource-intensive and the limit for such requests has been
+      // exceeded, retrieve the request ID and add an error response
+      final Integer requestId = rpcRequest.getInteger("id", null);
```

Comment:
will be nice to have a metrics for this kind of error, if not already present

Reply:
agree metrics would be nice - could be a separate PR though